### PR TITLE
Add Trino 466 release notes

### PR DESCRIPTION
**Pull request by mosabua**
## Description

Assemble the release notes for the upcoming Trino 466 release.

## Additional context and related issues

See https://github.com/trinodb/trino/pulls?q=is%3Apr+is%3Aclosed+milestone%3A466

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.

## Verification for each pull request

Format: PR/issue number, ✅ / ❌ rn ✅ / ❌ docs
✅ rn - release note added and verified, or assessed to be not necessary, set to ❌ rn before completion
✅ docs - need for docs assessed and merged, or assessed to be not necessary, set to ❌ docs before completion

## 21 Nov 2024

* #23812 ✅ rn ✅ docs
* #24211 ✅ rn ✅ docs
* #24193 ✅ rn ✅ docs
* #24214 ✅ rn ✅ docs, #23836
* #24215 ✅ rn ✅ docs
* #20555 ✅ rn ✅ docs
* #24223 ✅ rn ✅ docs

## 22 Nov 2024

* #24226 ✅ rn ✅ docs
* #24110 ❌ rn ✅ docs
* #24231 ✅ rn ✅ docs
* #24159 ❌ rn ✅ docs
* #24233 ✅ rn ✅ docs
* #24075 ✅ rn ✅ docs
* #24232 ✅ rn ✅ docs
* #24221 ✅ rn ✅ docs

## 23 Nov 2024

* #22675 ✅ rn ✅ docs
* #24237 ✅ rn ✅ docs

## 24 Nov 2024

* #24238 ✅ rn ✅ docs

## 25 Nov 2024

* #24239 ✅ rn ✅ docs
* #24245 ✅ rn ✅ docs
* #24246 ✅ rn ✅ docs
* #24191 ✅ rn ✅ docs
* #24252 ✅ rn ✅ docs
* #24254 ✅ rn ✅ docs

## 26 Nov 2024

* #24257 ✅ rn ✅ docs
* #24021 ✅ rn ❌ docs, #24285
* #24260 ✅ rn ✅ docs
* #24267 ✅ rn ✅ docs
* #24212 ✅ rn ✅ docs
* #24248 ❌ rn ❌ docs

## 27 Nov 2024

* #24271 ✅ rn ✅ docs
* #24262 ✅ rn ✅ docs
* #24273 ✅ rn ✅ docs
* #24270 ✅ rn ✅ docs
* #24243 ✅ rn ✅ docs
* #23002 ✅ rn ✅ docs
* #24276 ✅ rn ✅ docs
* #24277 ✅ rn ✅ docs
* #24263 ✅ rn ✅ docs
* #24268 ✅ rn ✅ docs
* #24283 ✅ rn ✅ docs
* #23836 ✅ rn ✅ docs

### CODE REVIEW

File: docs/src/main/sphinx/release/release-466.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# Release 466 (27 Nov 2024)
+
+## General
+
+* Improve performance of queries with grouping on joins. ({issue}`23812`)
+* TBD Spooling protocol is stable. Link to docs! ({issue}`24214`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:6]: fyi @wendigo .. will link to docs once I got that PR updated and merged with your help... and also reword this entry

File: docs/src/main/sphinx/release/release-466.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+* TBD Improve performance of querying `system.jdbc.tables` for multiple catalogs. ({issue}`24159`)
+* Support partial update. ({issue}`24075`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:8]: @electrum @djsstarburst @chenjian2664 .. do we somehow reword this to be more clear what it actually adds? Or leave it out and just announce the merge support in the connectors when the related PR is merged?

File: docs/src/main/sphinx/release/release-466.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+## Security
+
+* Add support [data access control with Apache
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:12]: @kokosing fyi

File: docs/src/main/sphinx/release/release-466.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+## Exasol connector
+
+## Faker connector
+
+## Google Sheets connector
+
+## Hive connector
+
+* TBD Improve performance of querying `system.jdbc.tables` for Hive, Iceberg, and Delta. ({issue}`24110`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:48]: Asked in original PR for clarification so I can reword .. also if it applies to Hudi

File: docs/src/main/sphinx/release/release-466.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+## MySQL connector
+
+## OpenSearch connector
+
+## Oracle connector
+
+## Phoenix connector
+
+* Support partial update. ({issue}`24075`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:82]: See entry in general

**Comment by djsstarburst** [comment]
Good point, @mosabua.  A better title for the change would be: "In SQL MERGE change only updated columns".

File: docs/src/main/sphinx/release/release-466.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
+# Release 466 (27 Nov 2024)
+
+## General
+
+* Improve performance of queries with grouping on joins. ({issue}`23812`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:5]: > [...] queries involving `GROUP BY` and joins.

File: docs/src/main/sphinx/release/release-466.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+* Improve client protocol throughput by introducing the spooling protocol.
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:6]: @wendigo .. once PR is merged I will link .. but otherwise.. good?

File: docs/src/main/sphinx/release/release-466.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  ({issue}`24214`)
+* Improve performance of querying `system.jdbc.tables` for multiple catalogs.
+  ({issue}`24159`)
+* Add support for connectors to improve performance for `MERGE` statements by
+  only changing updated columns. ({issue}`24075`)
+* Add support for changing field types inside a map. ({issue}`24248`)
+* Replace Python-based launcher script with fully-compatible, go-based
+  executable. Removes the requirement for a Python installation on Trino cluster
+  nodes. ({issue}`24271`)
+* Improve error messages for `INTERVAL` literals. ({issue}`23002`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:16]: Typically we dont add that.. wdyt @martint ?

File: docs/src/main/sphinx/release/release-466.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+* Improve performance of queries with grouping on joins. ({issue}`23812`)
+* Improve client protocol throughput by introducing the spooling protocol.
+  ({issue}`24214`)
+* Improve performance of querying `system.jdbc.tables` for multiple catalogs.
+  ({issue}`24159`)
+* Add support for connectors to improve performance for `MERGE` statements by
+  only changing updated columns. ({issue}`24075`)
+* Add support for changing field types inside a map. ({issue}`24248`)
+* Replace Python-based launcher script with fully-compatible, go-based
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:13]: @wendigo ?

File: docs/src/main/sphinx/release/release-466.md (hunk 7 above)
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:6]: There's no user visible change other than renaming some properties. The new protocol has been in Trino for a few releases now.

File: docs/src/main/sphinx/release/release-466.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
+# Release 466 (27 Nov 2024)
+
+## General
+
+* Improve performance of queries with grouping on joins. ({issue}`23812`)
+* Improve client protocol throughput by introducing the spooling protocol.
+  ({issue}`24214`)
+* Improve performance of querying `system.jdbc.tables` for multiple catalogs.
+  ({issue}`24159`)
+* Add support for connectors to improve performance for `MERGE` statements by
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:10]: @djsstarburst .. not sure we want to add this here .. I dont think we typically do this .. @martint ?

File: docs/src/main/sphinx/release/release-466.md (hunk 6 above)
Comment (colebow) [inline docs/src/main/sphinx/release/release-466.md:5]: ``with `GROUP BY` on the result of a join``

File: docs/src/main/sphinx/release/release-466.md (hunk 7 above)
Comment (colebow) [reply docs/src/main/sphinx/release/release-466.md:6]: If we're introducing a new protocol, we should probably say add, not improve.
``Add support for spooling client protocol, which...``

File: docs/src/main/sphinx/release/release-466.md (hunk 10 above)
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:10]: This should go in the SPI section, if there are SPI changes.

If it improves performance for any existing connectors, it should be added there, too.

The "by only changing updated columns" is an implementation detail we can omit here.

File: docs/src/main/sphinx/release/release-466.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  only changing updated columns. ({issue}`24075`)
+* Add support for changing field types inside a map. ({issue}`24248`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:12]: It's not clear what this means. In what context is this change allowed? What exactly does changing a field type inside a map mean?

File: docs/src/main/sphinx/release/release-466.md (hunk 9 above)
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:13]: There's no user visible impact. This is just an implementation detail. Remove it.

File: docs/src/main/sphinx/release/release-466.md (hunk 8 above)
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:16]: Remove this entry. We don't document improvements to error messages, as they are self-evident.

File: docs/src/main/sphinx/release/release-466.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+## Security
+
+* Add support [data access control with Apache
+  Ranger](/security/apache-ranger-access-control), including support for
+  column-masking, row-filtering, and audit logging. ({issue}`22675`)
+
+## JDBC driver
+
+* Automatically use the spooling protocol, when configured on the Trino cluster,
```
Comment (colebow) [inline docs/src/main/sphinx/release/release-466.md:26]:
```suggestion
* Automatically use the spooling protocol when it is configured on the Trino cluster,
```
Same for all future versions of this note, I think it's clearer

Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:26]: @wendigo

File: docs/src/main/sphinx/release/release-466.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  and add the parameter `encoding` to optionally set the preferred encoding from
+  the JDBC driver. ({issue}`24214`)
+* Improve performance of decompression when using Java 22 or newer as runtime
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:29]: @wendigo

File: docs/src/main/sphinx/release/release-466.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+  for the JDBC driver. ({issue}`24263`)
+
+## Docker image
+
+* Replace Python-based launcher script with fully-compatible, go-based
```
Comment (colebow) [inline docs/src/main/sphinx/release/release-466.md:34]: Is there a user impact for this? Or does the Docker image work just the same with this change being under the hood?

File: docs/src/main/sphinx/release/release-466.md (hunk 3 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-466.md:20]: "Add support **for** ..."?

column-masking -> column masking
row-filtering -> row filtering

File: docs/src/main/sphinx/release/release-466.md (hunk 10 above)
Comment (colebow) [inline docs/src/main/sphinx/release/release-466.md:10]: Remember to sort new features -> improvements -> fixes

File: docs/src/main/sphinx/release/release-466.md (hunk 13 above)
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:29]: Improve decompression performance when running the client with Java 22 or newer.

File: docs/src/main/sphinx/release/release-466.md (hunk 12 above)
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:26]: What's the user visible impact?

File: docs/src/main/sphinx/release/release-466.md [hunk 15]
```diff
@@ -0,0 +1,80 @@
+# Release 466 (27 Nov 2024)
+
+## General
+
+* Improve performance of queries with grouping on joins. ({issue}`23812`)
+* Improve client protocol throughput by introducing the spooling protocol.
+  ({issue}`24214`)
+* Improve performance of querying `system.jdbc.tables` for multiple catalogs.
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:8]: Need clarification here @piotrrzysko @raunaqmorarka @Praveen2112 @hashhar .. what is the user facing impact here .. when and where is the system.jdbc.tables info accessed?

File: docs/src/main/sphinx/release/release-466.md (hunk 14 above)
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:34]: This doesn't need to be documented. It's an implementation detail.

File: docs/src/main/sphinx/release/release-466.md [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+* Improve performance of querying `system.jdbc.tables`. ({issue}`24110`)
+
+## Iceberg connector
+
+* Add support for the [object store file
+  layout](https://iceberg.apache.org/docs/latest/aws/#object-store-file-layout).
+  ({issue}`8861`)
+* Add support for changing field types inside a map. ({issue}`24248`)
+* Improve performance of queries with selective joins. ({issue}`24277`)
+* Improve performance of querying `system.jdbc.tables`. ({issue}`24110`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-466.md:64]: Need clarification here @piotrrzysko @raunaqmorarka @Praveen2112 @hashhar .. what is the user facing impact here .. when and where is the system.jdbc.tables info accessed?

File: docs/src/main/sphinx/release/release-466.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+* Replace Python-based launcher script with fully-compatible, go-based
+  executable, and remove Python from the Docker image. ({issue}`24271`)
+
+## CLI
+
+* Automatically use the spooling protocol, when configured on the Trino cluster,
+  and add the option `--encoding` to optionally set the preferred encoding from
+  the CLI. ({issue}`24214`)
+* Improve performance of decompression when using Java 22 or newer as runtime
+  for the CLI. ({issue}`24263`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:43]: See above.

File: docs/src/main/sphinx/release/release-466.md [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+* Improve performance of decompression when using Java 22 or newer as runtime
+  for the CLI. ({issue}`24263`)  
+
+## BigQuery connector
+
+* Add support for `LIMIT` pushdown. ({issue}`23937`)
+
+## Delta Lake connector
+
+* Improve performance of querying `system.jdbc.tables`. ({issue}`24110`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:51]: This sounds weird. How about "Improve performance **when** querying ..."

File: docs/src/main/sphinx/release/release-466.md [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+## Iceberg connector
+
+* Add support for the [object store file
+  layout](https://iceberg.apache.org/docs/latest/aws/#object-store-file-layout).
+  ({issue}`8861`)
+* Add support for changing field types inside a map. ({issue}`24248`)
+* Improve performance of queries with selective joins. ({issue}`24277`)
+* Improve performance of querying `system.jdbc.tables`. ({issue}`24110`)
+* Fix failure when reading tables where a nested field has schema changes on
```
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:65]: What does "nested field has schema changes" mean?

File: docs/src/main/sphinx/release/release-466.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+  Parquet files. ({issue}`22922`)
+
+## Phoenix connector
+
+* Improve performance for `MERGE` statements by only changing updated columns. ({issue}`24075`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:70]: Remove the "by only changing updated columns" part. That's an implementation detail.

**Review by martint — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/release/release-466.md (hunk 14 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:34]: Well.. it is smaller and has less CVEs and some people get into the container and try to run Python stuff ..

Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:34]: Technically it should be transparent but nevertheless I think we have to mention it.

File: docs/src/main/sphinx/release/release-466.md (hunk 10 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:10]: Yeah .. that one probably has to go

File: docs/src/main/sphinx/release/release-466.md (hunk 7 above)
Comment (wendigo) [reply docs/src/main/sphinx/release/release-466.md:6]: @martint yes, but wasn't called out and documented

File: docs/src/main/sphinx/release/release-466.md (hunk 14 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-466.md:34]: Then we should just say "Remove Python from the Docker image".

File: docs/src/main/sphinx/release/release-466.md (hunk 7 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:6]: Yeah .. and we never announced it or added anything in any release notes.. we were waiting until docs are in and it is stable.. that is now.. cc @wendigo

File: docs/src/main/sphinx/release/release-466.md (hunk 18 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-466.md:51]: True. The impact is actually in the JDBC client. These are tables that the client uses to obtain metadata to implement some of the APIs from JDBC.

File: docs/src/main/sphinx/release/release-466.md (hunk 7 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:6]: Its technically an extension and fully compatible so its the same protocol still..

Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:6]: Also the user impact is mostly the throughput improvement.. it really only make sense fully if we merge the docs PR.

File: docs/src/main/sphinx/release/release-466.md (hunk 12 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:26]: Added

File: docs/src/main/sphinx/release/release-466.md (hunk 2 above)
Comment (hashhar) [reply docs/src/main/sphinx/release/release-466.md]: this is not user-visible change, it allows better planning of MERGE queries in JDBC connectors as far as I understand

Please confirm with @djsstarburst and @electrum though.

File: docs/src/main/sphinx/release/release-466.md (hunk 4 above)
Comment (hashhar) [reply docs/src/main/sphinx/release/release-466.md]: Not to hudi.

File: docs/src/main/sphinx/release/release-466.md (hunk 7 above)
Comment (hashhar) [reply docs/src/main/sphinx/release/release-466.md:6]: it's new protocol since it requires new clients for users to be able to make use of it.

it's forward compatible though (i.e. old clients can still access new servers)

File: docs/src/main/sphinx/release/release-466.md (hunk 8 above)
Comment (hashhar) [reply docs/src/main/sphinx/release/release-466.md:16]: yes, I think so too but was unsure.

File: docs/src/main/sphinx/release/release-466.md (hunk 9 above)
Comment (hashhar) [reply docs/src/main/sphinx/release/release-466.md:13]: yes.

File: docs/src/main/sphinx/release/release-466.md (hunk 18 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:51]: What JDBC client? Also in what case is the jdbc client even used .. the metadata sits in teh metastore .. its the a jdbc abstraction on top that applies to all metastore/catalogs?

File: docs/src/main/sphinx/release/release-466.md (hunk 15 above)
Comment (hashhar) [reply docs/src/main/sphinx/release/release-466.md:8]: Replied on the original PR already.

People can query it on their own to see metadata. BI tools for sure use it to inspect metadata.

File: docs/src/main/sphinx/release/release-466.md (hunk 18 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:48]: Got it now ..

File: docs/src/main/sphinx/release/release-466.md (hunk 15 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:10]: Great.. updated it now.

File: docs/src/main/sphinx/release/release-466.md (hunk 2 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md]: Clarified and updated already.

File: docs/src/main/sphinx/release/release-466.md (hunk 19 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-466.md:55]: I think this is related to the other change about nested fields

File: docs/src/main/sphinx/release/release-466.md (hunk 18 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-466.md]: JDBC client == Trino JDBC driver. This table is used in the implementation of `java.sql.DatabaseMetaData.getTables()`

File: docs/src/main/sphinx/release/release-466.md [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+## JDBC driver
+
+* Improve throughput by automatically using the [spooling
+  protocol](jdbc-spooling-protocol) when it is configured on the Trino cluster,
+  and add the parameter `encoding` to optionally set the preferred encoding from
+  the JDBC driver. ({issue}`24214`)
+* Improve decompression performance when running the client with Java 22 or
+  newer. ({issue}`24263`)
+* Improve performance of querying catalog metadata from `system.jdbc.tables`.
```
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:26]: Instead of mentioning the underlying table (which is an implementation detail of the driver), describe which API is now faster.

File: docs/src/main/sphinx/release/release-466.md (hunk 19 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-466.md:56]: The grammar seems off, too. How about this, assuming I'm interpreting the entry correctly:

> Fix failure when reading reading columns containing nested `row` types that differ from the schema of the underlying Parquet data.

File: docs/src/main/sphinx/release/release-466.md [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+## SQL Server connector
+
+* Rename the `sqlserver.experimental.stored-procedure-table-function-enabled`
+  configuration property to `sqlserver.stored-procedure-table-function-enabled`.
+  ({issue}`24239`)
+
+## SPI
+
+* Add `ConnectorSplit` argument to the `SystemTable.cursor()` method. ({issue}`24159`)
+* Add support for connectors to improve performance for `MERGE` statements. ({issue}`24075`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-466.md:72]: > Add support for partial row updates to `ConnectorMetadata.beginMerge()`.

**Review by martint — APPROVED** [review]
